components/esm/esmcommon.hpp

```diff
@@ -89,28 +93,41 @@ struct FIXED_STRING : public FIXED_STRING_BASE<FIXED_STRING, SIZE>
 
 // In the case of SIZE=4, it can be more efficient to match the string
 // as a 32 bit number, therefore the struct is implemented as a union with an int.
-template <>
-struct FIXED_STRING<4> : public FIXED_STRING_BASE<FIXED_STRING, 4>
+struct NAME
 {
-    union {
-        char data[4];
+    enum { size = 4 };
+
+    union
+    {
+        char data[size];
         uint32_t intval;
     };
 
-    using FIXED_STRING_BASE::operator==;
-    using FIXED_STRING_BASE::operator!=;
-
     bool operator==(uint32_t v) const { return v == intval; }
     bool operator!=(uint32_t v) const { return v != intval; }
+    bool operator==(const char* const str) const
+    {
+        for(size_t i = 0; i < size; ++i)
+        {
+            if(data[i] != str[i]) return false;
+            else if(data[i] == '\0') return true;
+        }
+        return str[size] == '\0';
+    }
+    bool operator!=(const char* const str) const { return !( (*this) == str ); }
+    bool operator==(const std::string& str) const { return (*this) == str.c_str(); }
+    bool operator!=(const std::string& str) const { return !( (*this) == str ); }
 
-    char const* ro_data() const { return data; }
-    char*       rw_data() { return data; }
+    char* rw_data() { return data; }
+
+    size_t data_size() const { return size; }
```

Comment:
Shouldn't this function be made static? Even if static isn't needed here, it expresses the intent more clearly.

Reply:
Done.